docs/DeveloperGuide.md

```diff
@@ -1,22 +1,264 @@
+
+3. **Methods**:
+   - Topic and Question Management:
+     - `parseTopic(String line)`: Parses a line from the questions file to add questions to the appropriate topic. Each line includes information such as topic name, question type, question text, and correct answer.
+     - `getOrCreateTopic(String topicName)`: Finds an existing topic by name or creates a new one if it doesn’t exist.
+     - `addFlashcardByUser(String input)`: Allows user to add a custom flashcard question and answer to the quiz system.
+   - Quiz Operations:
+     - `selectTopic(String topicName)`: Allows the user to choose a topic for the quiz. If the topic exists, a new quiz session is started.
+     - `startQuiz(Topic topic, Scanner quizScanner)`: Begins a quiz session for the specified topic, sets up the timer and question limit, and calculates the score upon completion.
+   - Utility Methods:
+     - `printTopics()`: Displays all available topics to the user.
+     - `addTopic(Topic topic)`: Adds a new topic to the topics list if it doesn’t already exist.
+     - `removeTopic(Topic topic)`: Removes an existing topic from the topics list.
+   - Data Persistence:
+     - `saveDataToFile()`: Saves past quiz results to the results file.
+     - `loadDataFromFile()`: Loads questions and past results from their respective files using `Storage`.
+   - Helper Methods:
+     - `addPastResult(int score, String comment)`: Adds the quiz score and a feedback comment to pastResults.
+     - `generateComment(int score)`: Provides feedback based on the user’s score.
+
+#### **Design Rationale**
+
+- **Modularization**: By separating topics, quiz sessions, and persistence into distinct methods and attributes, the class becomes more manageable and modular. Each component has a clear responsibility, which improves readability and maintainability.
+- **Scalability**: Storing topics and questions in collections allows the quiz system to scale easily. New topics and questions can be added without modifying core functionality.
+- **Separation of Concerns**: The QuizManager focuses solely on managing quiz-related data and processes, delegating storage tasks to the Storage class. This separation makes the code more maintainable and allows each component to evolve independently.
+  
+
+#### **Alternative Considerations**
+
+1. **Encapsulate Score Calculation**:
+   - Alternative: Create separate classes for `ScoreCalculator`.
+      - Pros: Increases code modularity and allows for more scoring logic.
+      - Cons: Adds complexity that may not be necessary if the scoring logic remain simple.
+
+#### UML Class Diagram
+Below is the class diagram for the `QuizManager` class, illustrating its attributes and methods.
+![image](https://github.com/user-attachments/assets/85b30191-76ab-431f-8cdb-0e29c481e506)
+
```

Comment:
Perhaps including a sequence diagram here would be more meaningful? The class diagram alone does not really give much insight into how QuizManager works.